stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,535 @@
+		}
+		l, err := net.Listen("tcp4", "127.0.0.1:0")
+		if err != nil {
+			log.PrintE("unable to open stdout listener", err)
+			os.Exit(254)
+		}
+		endpoints[1] = &l
+		port := strconv.Itoa((*endpoints[1]).Addr().(*net.TCPAddr).Port)
+		ep := strings.Join([]string{"stdout", "AF_INET4", "127.0.0.1", port}, ",") + "\n"
+		_, _ = statusFile.Write([]byte(ep))
+		diag.Printf("Listening for stdout on %s", port)
+
+	}
+	if stderrMode == "true" {
+		fifos[2], err = os.OpenFile(filepath.Join(pathPrefix, appName, "stage2-stderr"), os.O_RDONLY, os.ModeNamedPipe)
+		if err != nil {
+			log.PrintE("invalid stderr FIFO", err)
+			os.Exit(254)
+		}
+		l, err := net.Listen("tcp4", "127.0.0.1:0")
+		if err != nil {
+			log.PrintE("unable to open stdout listener", err)
+			os.Exit(254)
+		}
+		endpoints[2] = &l
+		port := strconv.Itoa((*endpoints[2]).Addr().(*net.TCPAddr).Port)
+		ep := strings.Join([]string{"stderr", "AF_INET4", "127.0.0.1", port}, ",") + "\n"
+		_, _ = statusFile.Write([]byte(ep))
+		diag.Printf("Listening for stderr on %s", port)
+
+	}
+	statusFile.Sync()
+	statusFile.Close()
+	c := make(chan error)
+
+	// Enabled by `RKT_EXPERIMENT_LOGMODE=k8s-plain`
+	// TODO(lucab): stabilize this and implement forwarding to journald
+	var logFile *os.File
+	switch logMode {
+	case "k8s-plain":
+		var err error
+		// TODO(lucab): check what should be the target path with Euan
+		logTarget := filepath.Join(pathPrefix, appName, "logfile")
+		logFile, err = os.OpenFile(logTarget, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	// proxy stdin
+	if fifos[0] != nil && endpoints[0] != nil {
+		clients := make(chan *net.Conn)
+		go acceptConn(endpoints[0], clients, "stdin")
+		go muxInput(clients, fifos[0])
+	}
+
+	// proxy stdout
+	if fifos[1] != nil && endpoints[1] != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan *net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(fifos[1], lines)
+		go acceptConn(endpoints[1], clients, "stdout")
+		go muxOutput("stdout", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// proxy stderr
+	if fifos[2] != nil && endpoints[2] != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan *net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(fifos[2], lines)
+		go acceptConn(endpoints[2], clients, "stderr")
+		go muxOutput("stderr", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	return <-c
+}
+
+// bufferLine buffers and queues a single line from a Reader to a multiplexer
+func bufferLine(r io.Reader, c chan []byte) {
+	rd := bufio.NewReader(r)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if err == nil {
+			c <- lineOut
+		}
+		// TODO(lucab): stop and signal caller on error
+	}
+}
+
+// acceptConn accepts a single client and queues it for further proxying
+func acceptConn(socket *net.Listener, c chan *net.Conn, stream string) {
```

Comment:
`net.Conn` is an interface, please don't send pointers of it via a channel, it boils down to being a two-word pair of a pointer and type already, see http://research.swtch.com/interfaces.